// shakapacker/package/utils/errorCodes.js
/**
 * Error codes for programmatic error handling in Shakapacker
 * These codes allow consumers to handle specific errors programmatically
 * @module shakapacker/utils/errorCodes
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShakapackerError = exports.ErrorMessages = exports.ErrorCode = void 0;
exports.createError = createError;
exports.isShakapackerError = isShakapackerError;
exports.getErrorCode = getErrorCode;
/**
 * Error code enumeration for Shakapacker errors
 */
var ErrorCode;
(function (ErrorCode) {
    // Configuration errors (1xxx)
    ErrorCode["CONFIG_NOT_FOUND"] = "SHAKAPACKER_1001";
    ErrorCode["CONFIG_INVALID_YAML"] = "SHAKAPACKER_1002";
    ErrorCode["CONFIG_MISSING_REQUIRED"] = "SHAKAPACKER_1003";
    ErrorCode["CONFIG_VALIDATION_FAILED"] = "SHAKAPACKER_1004";
    ErrorCode["CONFIG_MERGE_FAILED"] = "SHAKAPACKER_1005";
    ErrorCode["CONFIG_TYPE_MISMATCH"] = "SHAKAPACKER_1006";
    // File system errors (2xxx)
    ErrorCode["FILE_NOT_FOUND"] = "SHAKAPACKER_2001";
    ErrorCode["FILE_READ_ERROR"] = "SHAKAPACKER_2002";
    ErrorCode["FILE_WRITE_ERROR"] = "SHAKAPACKER_2003";
    ErrorCode["FILE_PERMISSION_DENIED"] = "SHAKAPACKER_2004";
    ErrorCode["PATH_TRAVERSAL_DETECTED"] = "SHAKAPACKER_2005";
    ErrorCode["INVALID_PATH"] = "SHAKAPACKER_2006";
    // Module errors (3xxx)
    ErrorCode["MODULE_NOT_FOUND"] = "SHAKAPACKER_3001";
    ErrorCode["MODULE_LOAD_FAILED"] = "SHAKAPACKER_3002";
    ErrorCode["MODULE_INVALID_EXPORT"] = "SHAKAPACKER_3003";
    ErrorCode["LOADER_NOT_FOUND"] = "SHAKAPACKER_3004";
    ErrorCode["PLUGIN_NOT_FOUND"] = "SHAKAPACKER_3005";
    ErrorCode["PLUGIN_INVALID"] = "SHAKAPACKER_3006";
    // Environment errors (4xxx)
    ErrorCode["ENV_INVALID_NODE_ENV"] = "SHAKAPACKER_4001";
    ErrorCode["ENV_MISSING_REQUIRED"] = "SHAKAPACKER_4002";
    ErrorCode["ENV_INVALID_VALUE"] = "SHAKAPACKER_4003";
    ErrorCode["ENV_SANITIZATION_REQUIRED"] = "SHAKAPACKER_4004";
    // Bundler errors (5xxx)
    ErrorCode["BUNDLER_UNSUPPORTED"] = "SHAKAPACKER_5001";
    ErrorCode["BUNDLER_CONFIG_INVALID"] = "SHAKAPACKER_5002";
    ErrorCode["WEBPACK_CONFIG_INVALID"] = "SHAKAPACKER_5003";
    ErrorCode["RSPACK_CONFIG_INVALID"] = "SHAKAPACKER_5004";
    ErrorCode["TRANSPILER_NOT_FOUND"] = "SHAKAPACKER_5005";
    ErrorCode["TRANSPILER_CONFIG_INVALID"] = "SHAKAPACKER_5006";
    // Dev server errors (6xxx)
    ErrorCode["DEVSERVER_CONFIG_INVALID"] = "SHAKAPACKER_6001";
    ErrorCode["DEVSERVER_PORT_INVALID"] = "SHAKAPACKER_6002";
    ErrorCode["DEVSERVER_PORT_IN_USE"] = "SHAKAPACKER_6003";
    ErrorCode["DEVSERVER_START_FAILED"] = "SHAKAPACKER_6004";
    // Security errors (7xxx)
    ErrorCode["SECURITY_PATH_TRAVERSAL"] = "SHAKAPACKER_7001";
    ErrorCode["SECURITY_INVALID_INPUT"] = "SHAKAPACKER_7002";
    ErrorCode["SECURITY_CONTROL_CHARS"] = "SHAKAPACKER_7003";
    ErrorCode["SECURITY_INJECTION_ATTEMPT"] = "SHAKAPACKER_7004";
    // Validation errors (8xxx)
    ErrorCode["VALIDATION_FAILED"] = "SHAKAPACKER_8001";
    ErrorCode["VALIDATION_TYPE_ERROR"] = "SHAKAPACKER_8002";
    ErrorCode["VALIDATION_RANGE_ERROR"] = "SHAKAPACKER_8003";
    ErrorCode["VALIDATION_FORMAT_ERROR"] = "SHAKAPACKER_8004";
    ErrorCode["VALIDATION_CONSTRAINT_ERROR"] = "SHAKAPACKER_8005";
    // Generic errors (9xxx)
    ErrorCode["UNKNOWN_ERROR"] = "SHAKAPACKER_9000";
    ErrorCode["INTERNAL_ERROR"] = "SHAKAPACKER_9001";
    ErrorCode["DEPRECATED_FEATURE"] = "SHAKAPACKER_9002";
    ErrorCode["NOT_IMPLEMENTED"] = "SHAKAPACKER_9003";
    ErrorCode["OPERATION_FAILED"] = "SHAKAPACKER_9004";
})(ErrorCode || (exports.ErrorCode = ErrorCode = {}));
/**
 * Error message templates for each error code
 */
exports.ErrorMessages = {
    // Configuration errors
    [ErrorCode.CONFIG_NOT_FOUND]: "Configuration file not found: {path}",
    [ErrorCode.CONFIG_INVALID_YAML]: "Invalid YAML in configuration file: {path}",
    [ErrorCode.CONFIG_MISSING_REQUIRED]: "Missing required configuration field: {field}",
    [ErrorCode.CONFIG_VALIDATION_FAILED]: "Configuration validation failed: {reason}",
    [ErrorCode.CONFIG_MERGE_FAILED]: "Failed to merge configurations: {reason}",
    [ErrorCode.CONFIG_TYPE_MISMATCH]: "Configuration type mismatch for {field}: expected {expected}, got {actual}",
    // File system errors
    [ErrorCode.FILE_NOT_FOUND]: "File not found: {path}",
    [ErrorCode.FILE_READ_ERROR]: "Error reading file: {path}",
    [ErrorCode.FILE_WRITE_ERROR]: "Error writing file: {path}",
    [ErrorCode.FILE_PERMISSION_DENIED]: "Permission denied accessing: {path}",
    [ErrorCode.PATH_TRAVERSAL_DETECTED]: "Path traversal attempt detected: {path}",
    [ErrorCode.INVALID_PATH]: "Invalid path: {path}",
    // Module errors
    [ErrorCode.MODULE_NOT_FOUND]: "Module not found: {module}",
    [ErrorCode.MODULE_LOAD_FAILED]: "Failed to load module: {module}",
    [ErrorCode.MODULE_INVALID_EXPORT]: "Invalid export from module: {module}",
    [ErrorCode.LOADER_NOT_FOUND]: "Loader not found: {loader}",
    [ErrorCode.PLUGIN_NOT_FOUND]: "Plugin not found: {plugin}",
    [ErrorCode.PLUGIN_INVALID]: "Invalid plugin: {plugin}",
    // Environment errors
    [ErrorCode.ENV_INVALID_NODE_ENV]: "Invalid NODE_ENV value: {value}. Valid values are: {valid}",
    [ErrorCode.ENV_MISSING_REQUIRED]: "Missing required environment variable: {variable}",
    [ErrorCode.ENV_INVALID_VALUE]: "Invalid value for environment variable {variable}: {value}",
    [ErrorCode.ENV_SANITIZATION_REQUIRED]: "Environment variable {variable} contained unsafe characters and was sanitized",
    // Bundler errors
    [ErrorCode.BUNDLER_UNSUPPORTED]: "Unsupported bundler: {bundler}",
    [ErrorCode.BUNDLER_CONFIG_INVALID]: "Invalid bundler configuration: {reason}",
    [ErrorCode.WEBPACK_CONFIG_INVALID]: "Invalid webpack configuration: {reason}",
    [ErrorCode.RSPACK_CONFIG_INVALID]: "Invalid rspack configuration: {reason}",
    [ErrorCode.TRANSPILER_NOT_FOUND]: "Transpiler not found: {transpiler}",
    [ErrorCode.TRANSPILER_CONFIG_INVALID]: "Invalid transpiler configuration: {reason}",
    // Dev server errors
    [ErrorCode.DEVSERVER_CONFIG_INVALID]: "Invalid dev server configuration: {reason}",
    [ErrorCode.DEVSERVER_PORT_INVALID]: "Invalid port: {port}",
    [ErrorCode.DEVSERVER_PORT_IN_USE]: "Port {port} is already in use",
    [ErrorCode.DEVSERVER_START_FAILED]: "Failed to start dev server: {reason}",
    // Security errors
    [ErrorCode.SECURITY_PATH_TRAVERSAL]: "Security: Path traversal attempt blocked: {path}",
    [ErrorCode.SECURITY_INVALID_INPUT]: "Security: Invalid input detected: {input}",
    [ErrorCode.SECURITY_CONTROL_CHARS]: "Security: Control characters detected and removed from: {field}",
    [ErrorCode.SECURITY_INJECTION_ATTEMPT]: "Security: Potential injection attempt blocked: {details}",
    // Validation errors
    [ErrorCode.VALIDATION_FAILED]: "Validation failed: {reason}",
    [ErrorCode.VALIDATION_TYPE_ERROR]: "Type validation error: {field} should be {type}",
    [ErrorCode.VALIDATION_RANGE_ERROR]: "Value out of range: {field} must be between {min} and {max}",
    [ErrorCode.VALIDATION_FORMAT_ERROR]: "Format error: {field} does not match expected format",
    [ErrorCode.VALIDATION_CONSTRAINT_ERROR]: "Constraint violation: {constraint}",
    // Generic errors
    [ErrorCode.UNKNOWN_ERROR]: "An unknown error occurred",
    [ErrorCode.INTERNAL_ERROR]: "Internal error: {details}",
    [ErrorCode.DEPRECATED_FEATURE]: "Deprecated feature: {feature}. Use {alternative} instead",
    [ErrorCode.NOT_IMPLEMENTED]: "Feature not yet implemented: {feature}",
    [ErrorCode.OPERATION_FAILED]: "Operation failed: {operation}"
};
/**
 * Shakapacker error class with error code support
 */
class ShakapackerError extends Error {
    constructor(code, details, customMessage) {
        const template = exports.ErrorMessages[code] || "An error occurred";
        const message = customMessage || ShakapackerError.formatMessage(template, details);
        super(message);
        this.name = "ShakapackerError";
        this.code = code;
        this.details = details;
        // Maintain proper stack trace for where error was thrown
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ShakapackerError);
        }
    }
    /**
     * Format error message with template values
     */
    static formatMessage(template, details) {
        if (!details)
            return template;
        return template.replace(/{(\w+)}/g, (match, key) => {
            const value = details[key];
            if (value === undefined)
                return match;
            if (typeof value === "object") {
                return JSON.stringify(value);
            }
            return String(value);
        });
    }
    /**
     * Convert error to JSON for logging or API responses
     */
    toJSON() {
        return {
            name: this.name,
            code: this.code,
            message: this.message,
            details: this.details,
            stack: this.stack
        };
    }
}
exports.ShakapackerError = ShakapackerError;
/**
 * Helper function to create a Shakapacker error
 */
function createError(code, details) {
    return new ShakapackerError(code, details);
}
/**
 * Helper function to check if an error is a Shakapacker error
 */
function isShakapackerError(error) {
    return error instanceof ShakapackerError;
}
/**
 * Helper function to get error code from any error
 */
function getErrorCode(error) {
    if (isShakapackerError(error)) {
        return error.code;
    }
    return null;
}
