// shakapacker/package/utils/errorHelpers.js
/**
 * Error handling utilities for consistent error management
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFileNotFoundError = isFileNotFoundError;
exports.isModuleNotFoundError = isModuleNotFoundError;
exports.createFileOperationError = createFileOperationError;
exports.createFileOperationErrorLegacy = createFileOperationErrorLegacy;
exports.getErrorMessage = getErrorMessage;
exports.isNodeError = isNodeError;
exports.createConfigValidationErrorWithCode = createConfigValidationErrorWithCode;
exports.createModuleNotFoundError = createModuleNotFoundError;
exports.createPathTraversalError = createPathTraversalError;
exports.createPortValidationError = createPortValidationError;
const errorCodes_1 = require("./errorCodes");
/**
 * Checks if an error is a file not found error (ENOENT)
 */
function isFileNotFoundError(error) {
    return (error !== null &&
        typeof error === "object" &&
        "code" in error &&
        error.code === "ENOENT");
}
/**
 * Checks if an error is a module not found error
 */
function isModuleNotFoundError(error) {
    return (error !== null &&
        typeof error === "object" &&
        "code" in error &&
        error.code === "MODULE_NOT_FOUND");
}
/**
 * Creates a consistent error message for file operations
 */
function createFileOperationError(operation, filePath, details) {
    let errorCode;
    if (operation === "read") {
        errorCode = errorCodes_1.ErrorCode.FILE_READ_ERROR;
    }
    else if (operation === "write") {
        errorCode = errorCodes_1.ErrorCode.FILE_WRITE_ERROR;
    }
    else {
        errorCode = errorCodes_1.ErrorCode.FILE_NOT_FOUND;
    }
    return new errorCodes_1.ShakapackerError(errorCode, {
        path: filePath,
        operation,
        details
    });
}
/**
 * Creates a consistent error message for file operations (backward compatibility)
 */
function createFileOperationErrorLegacy(operation, filePath, details) {
    const baseMessage = `Failed to ${operation} file at path '${filePath}'`;
    const errorDetails = details ? ` - ${details}` : "";
    let suggestion;
    if (operation === "read") {
        suggestion = " (check if file exists and permissions are correct)";
    }
    else if (operation === "write") {
        suggestion = " (check write permissions and disk space)";
    }
    else {
        suggestion = " (check permissions)";
    }
    return new Error(`${baseMessage}${errorDetails}${suggestion}`);
}
/**
 * Safely gets error message from unknown error type
 */
function getErrorMessage(error) {
    if (error instanceof Error) {
        // Include stack trace for better debugging in development
        const isDev = process.env.NODE_ENV === "development";
        return isDev && error.stack
            ? `${error.message}\n${error.stack}`
            : error.message;
    }
    if (typeof error === "string") {
        return error;
    }
    if (error && typeof error === "object" && "message" in error) {
        return String(error.message);
    }
    // Provide more context for truly unknown errors
    return `Unknown error occurred (type: ${typeof error}, value: ${JSON.stringify(error)})`;
}
/**
 * Type guard for NodeJS errors with errno
 */
function isNodeError(error) {
    return (error instanceof Error &&
        "code" in error &&
        typeof error.code === "string");
}
/**
 * Creates a configuration validation error
 */
function createConfigValidationErrorWithCode(configPath, environment, reason) {
    return new errorCodes_1.ShakapackerError(errorCodes_1.ErrorCode.CONFIG_VALIDATION_FAILED, {
        path: configPath,
        environment,
        reason
    });
}
/**
 * Creates a module not found error
 */
function createModuleNotFoundError(moduleName, details) {
    return new errorCodes_1.ShakapackerError(errorCodes_1.ErrorCode.MODULE_NOT_FOUND, {
        module: moduleName,
        details
    });
}
/**
 * Creates a path traversal security error
 */
function createPathTraversalError(path) {
    return new errorCodes_1.ShakapackerError(errorCodes_1.ErrorCode.SECURITY_PATH_TRAVERSAL, {
        path
    });
}
/**
 * Creates a port validation error
 */
function createPortValidationError(port) {
    return new errorCodes_1.ShakapackerError(errorCodes_1.ErrorCode.DEVSERVER_PORT_INVALID, {
        port: String(port)
    });
}
