// shakapacker/package/utils/helpers.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolvedPath = exports.packageMajorVersion = exports.packageFullVersion = exports.loaderMatches = exports.moduleExists = exports.canProcess = exports.ensureTrailingSlash = exports.isBoolean = void 0;
const { isModuleNotFoundError, getErrorMessage } = require("./errorHelpers");
const isBoolean = (str) => /^true/.test(str) || /^false/.test(str);
exports.isBoolean = isBoolean;
const ensureTrailingSlash = (path) => path.endsWith("/") ? path : `${path}/`;
exports.ensureTrailingSlash = ensureTrailingSlash;
const resolvedPath = (packageName) => {
    try {
        return require.resolve(packageName);
    }
    catch (error) {
        if (!isModuleNotFoundError(error)) {
            throw error;
        }
        return null;
    }
};
exports.resolvedPath = resolvedPath;
const moduleExists = (packageName) => !!resolvedPath(packageName);
exports.moduleExists = moduleExists;
const canProcess = (rule, fn) => {
    const modulePath = resolvedPath(rule);
    if (modulePath) {
        return fn(modulePath);
    }
    return null;
};
exports.canProcess = canProcess;
const loaderMatches = (configLoader, loaderToCheck, fn) => {
    // If transpiler is set to 'none', skip all transpiler rules (for custom webpack configs)
    if (configLoader === "none") {
        return null;
    }
    if (configLoader !== loaderToCheck) {
        return null;
    }
    const loaderName = `${configLoader}-loader`;
    if (!moduleExists(loaderName)) {
        throw new Error(`Your Shakapacker config specified using ${configLoader}, but ${loaderName} package is not installed.\n` +
            `\nTo fix this issue, run one of the following commands:\n` +
            `  npm install --save-dev ${loaderName}\n` +
            `  yarn add --dev ${loaderName}\n` +
            `\nOr change your 'javascript_transpiler' setting in shakapacker.yml to use a different loader.`);
    }
    return fn();
};
exports.loaderMatches = loaderMatches;
const packageFullVersion = (packageName) => {
    try {
        const packageJsonPath = require.resolve(`${packageName}/package.json`);
        // eslint-disable-next-line import/no-dynamic-require
        const packageJson = require(packageJsonPath);
        return packageJson.version;
    }
    catch (error) {
        // Re-throw the error with proper code to maintain compatibility with babel preset
        // The preset expects MODULE_NOT_FOUND errors to handle missing core-js gracefully
        if (error.code === "MODULE_NOT_FOUND") {
            throw error;
        }
        // For other errors, warn and re-throw
        console.warn(`[SHAKAPACKER WARNING] Failed to get version for package ${packageName}: ${getErrorMessage(error)}`);
        throw error;
    }
};
exports.packageFullVersion = packageFullVersion;
const packageMajorVersion = (packageName) => {
    const match = packageFullVersion(packageName).match(/^\d+/);
    return match ? parseInt(match[0], 10) : 0;
};
exports.packageMajorVersion = packageMajorVersion;
