// shakapacker/package/swc/index.js
/* eslint import/no-dynamic-require: 0 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSwcLoaderConfig = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
const webpack_merge_1 = require("webpack-merge");
const JSX_FILE_REGEX = /\.(jsx|tsx)(\.erb)?$/;
const TYPESCRIPT_FILE_REGEX = /\.(ts|tsx)(\.erb)?$/;
const isJsxFile = (filename) => !!filename.match(JSX_FILE_REGEX);
const isTypescriptFile = (filename) => !!filename.match(TYPESCRIPT_FILE_REGEX);
const getCustomConfig = () => {
    const path = (0, path_1.resolve)("config", "swc.config.js");
    if ((0, fs_1.existsSync)(path)) {
        return require(path);
    }
    return {};
};
const getSwcLoaderConfig = (filenameToProcess) => {
    const customConfig = getCustomConfig();
    const isTs = isTypescriptFile(filenameToProcess);
    const isJsx = isJsxFile(filenameToProcess);
    const jsxKey = isTs ? "tsx" : "jsx";
    const defaultConfig = {
        loader: require.resolve("swc-loader"),
        options: {
            jsc: {
                parser: {
                    dynamicImport: true,
                    syntax: isTs ? "typescript" : "ecmascript",
                    [jsxKey]: isJsx
                },
                loose: false
            },
            sourceMaps: true,
            env: {
                coreJs: 3,
                exclude: ["transform-typeof-symbol"],
                mode: "entry"
            }
        }
    };
    return (0, webpack_merge_1.merge)(defaultConfig, customConfig);
};
exports.getSwcLoaderConfig = getSwcLoaderConfig;
