// @kaspernj/api-maker/build/inputs/checkboxes.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable sort-imports */
import * as inflection from "inflection";
import { digs } from "diggerize";
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import BaseComponent from "../base-component";
import InvalidFeedback from "../bootstrap/invalid-feedback";
import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames"; // eslint-disable-line import/no-unresolved
import inputWrapper from "./input-wrapper";
import memo from "set-state-compare/build/memo.js";
import propTypesExact from "prop-types-exact";
const ApiMakerInputsCheckboxes = memo(shapeComponent(class ApiMakerInputsCheckboxes extends BaseComponent {
    static propTypes = propTypesExact({
        attribute: PropTypes.string,
        defaultValue: PropTypes.array,
        id: PropTypes.string,
        inputProps: PropTypes.object.isRequired,
        label: PropTypes.any,
        model: PropTypes.object,
        name: PropTypes.string,
        onChange: PropTypes.func,
        options: PropTypes.array.isRequired,
        wrapperOpts: PropTypes.object
    });
    render() {
        const { className, inputProps, label, wrapperOpts, ...restProps } = this.props;
        return (_jsxs("div", { className: classNames("component-bootstrap-check-boxes", className), ...restProps, children: [_jsx("input", { name: this.inputName(), ref: this.props.inputProps.ref, type: "hidden", value: "" }), this.props.options.map((option, index) => this.optionElement(option, index))] }));
    }
    inputDefaultValue() {
        const { attribute, defaultValue, model } = this.props;
        if (defaultValue) {
            return defaultValue;
        }
        else if (attribute && model) {
            if (!model[attribute]) {
                throw `No such attribute: ${attribute}`; // eslint-disable-line no-throw-literal
            }
            return this.props.model[attribute]();
        }
    }
    inputCheckboxClassName() {
        const classNames = [];
        if (this.props.wrapperOpts.errors.length > 0)
            classNames.push("is-invalid");
        return classNames.join(" ");
    }
    inputName() {
        if (this.props.name) {
            return `${this.props.name}[]`;
        }
        else if (this.props.model) {
            return `${this.props.model.modelClassData().paramKey}[${inflection.underscore(this.props.attribute)}]`;
        }
    }
    isDefaultSelected(option) {
        let defaultValue = this.inputDefaultValue();
        if (!defaultValue)
            return false;
        if (defaultValue.constructor === Array) {
            return defaultValue.includes(option);
        }
        else {
            return defaultValue == option;
        }
    }
    generatedId() {
        if (!this.generatedIdValue) {
            this.generatedIdValue = Math.random()
                .toString(36)
                .substring(2, 15) + Math.random()
                .toString(36)
                .substring(2, 15);
        }
        return this.generatedIdValue;
    }
    optionElement(option, index) {
        const { onChange, options, wrapperOpts } = this.props;
        const { errors } = digs(wrapperOpts, "errors");
        const id = `${this.generatedId()}-${index}`;
        return (_jsxs("div", { className: "checkboxes-option", children: [_jsx("input", { className: this.inputCheckboxClassName(), "data-option-value": option[1], defaultChecked: this.isDefaultSelected(option[1]), id: id, name: this.inputName(), onChange: onChange, type: "checkbox", value: option[1] }), _jsx("label", { className: "ml-1", htmlFor: id, children: option[0] }), index + 1 == options.length && errors.length > 0 &&
                    _jsx(InvalidFeedback, { errors: errors })] }, `option-${option[1]}`));
    }
}));
export default inputWrapper(ApiMakerInputsCheckboxes);
