// @kaspernj/api-maker/build/inputs/select.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable sort-imports */
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import { useForm } from "../form";
import BaseComponent from "../base-component";
import PropTypes from "prop-types";
import React from "react";
import inputWrapper from "./input-wrapper";
import memo from "set-state-compare/build/memo.js";
const ApiMakerInputsSelect = memo(shapeComponent(class ApiMakerInputsSelect extends BaseComponent {
    static propTypes = {
        attribute: PropTypes.string,
        children: PropTypes.any,
        defaultValue: PropTypes.oneOfType([PropTypes.array, PropTypes.number, PropTypes.string]),
        id: PropTypes.string,
        includeBlank: PropTypes.oneOfType([PropTypes.bool, PropTypes.any]),
        inputProps: PropTypes.object.isRequired,
        model: PropTypes.object,
        name: PropTypes.string,
        onChange: PropTypes.func,
        options: PropTypes.array,
        wrapperOpts: PropTypes.object.isRequired
    };
    setup() {
        this.form = useForm();
    }
    render() {
        const { attribute, children, defaultValue, id, includeBlank, inputProps, inputRef, model, name, onChange, options, wrapperOpts, ...restProps } = this.props;
        return (_jsxs("select", { onChange: this.tt.onChange, ...inputProps, ...restProps, children: [this.includeBlank() &&
                    _jsx("option", { "data-include-blank": "true", children: typeof includeBlank == "boolean" ? null : includeBlank }), options && options.map((option) => ( // eslint-disable-line no-extra-parens
                _jsx("option", { value: this.optionValue(option), children: this.optionLabel(option) }, this.optionKey(option)))), children] }));
    }
    onChange = (e) => {
        const { form } = this.tt;
        const { name, onChange } = this.props;
        if (form && name)
            form.setValue(name, e.target.value);
        if (onChange)
            onChange(e);
    };
    optionKey(option) {
        if (Array.isArray(option)) {
            return `select-option-${option[1]}`;
        }
        else {
            return `select-option-${option}`;
        }
    }
    optionLabel(option) {
        if (Array.isArray(option)) {
            return option[0];
        }
        else {
            return option;
        }
    }
    optionValue(option) {
        if (Array.isArray(option)) {
            return option[1];
        }
        else {
            return option;
        }
    }
    includeBlank() {
        return Boolean(this.props.includeBlank && !this.props.multiple);
    }
}));
export { ApiMakerInputsSelect as Select };
export default inputWrapper(ApiMakerInputsSelect);
