// @kaspernj/api-maker/build/modal.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable sort-imports */
import { Modal, Pressable, View } from "react-native";
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import BaseComponent from "./base-component";
import React from "react";
import memo from "set-state-compare/build/memo.js";
export default memo(shapeComponent(class ApiMakerModal extends BaseComponent {
    render() {
        const { children, onRequestClose, ...restProps } = this.props;
        return (_jsx(Modal, { onRequestClose: onRequestClose, ...restProps, children: _jsxs(View, { style: this.cache("rootViewStyle", {
                    alignItems: "center",
                    justifyContent: "center",
                    minWidth: "100%",
                    minHeight: "100%",
                    padding: 20
                }), children: [_jsx(Pressable, { dataSet: this.cache("pressableDataSet", { class: "modal-backdrop" }), onPress: onRequestClose, style: this.cache("pressableStyle", {
                            position: "absolute",
                            minWidth: "100%",
                            minHeight: "100%",
                            backgroundColor: "rgba(0, 0, 0, 0.5)"
                        }) }), children] }) }));
    }
}));
