// @kaspernj/api-maker/build/table/header-select.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable implicit-arrow-linebreak, react/jsx-sort-props, sort-imports */
import { Pressable, View } from "react-native";
import BaseComponent from "../base-component";
import Collection from "../collection.js";
import HeaderColumnContent from "./header-column-content";
import memo from "set-state-compare/build/memo.js";
import Modal from "../utils/modal";
import PropTypes from "prop-types";
import propTypesExact from "prop-types-exact";
import React from "react";
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import Text from "../utils/text";
import useI18n from "i18n-on-steroids/src/use-i18n.mjs";
import useSorting from "./use-sorting.js";
export default memo(shapeComponent(class ApiMakerTableHeaderSelect extends BaseComponent {
    static propTypes = propTypesExact({
        preparedColumns: PropTypes.array.isRequired,
        query: PropTypes.instanceOf(Collection).isRequired,
        table: PropTypes.object
    });
    setup() {
        const { t } = useI18n({ namespace: "js.api_maker.table.header_select" });
        const { preparedColumns, query } = this.p;
        const { qParams, searchKey, sortAttribute, sortMode } = useSorting({ query });
        this.useStates({
            modalOpen: false
        });
        this.setInstance({ qParams, searchKey, sortAttribute, sortMode, t });
        this.sortedByPreparedColumn = preparedColumns.find((preparedColumn) => preparedColumn.tableSettingColumn.sortKey() == sortAttribute);
    }
    render() {
        const { sortedByPreparedColumn, sortMode, t } = this.tt;
        const { table } = this.p;
        const column = sortedByPreparedColumn?.column;
        const columnLabel = column && table.headerLabelForColumn(column);
        return (_jsxs(View, { dataSet: this.cache("rootViewDataSet", { component: "api-maker/table/header-select" }), children: [this.s.modalOpen &&
                    _jsx(Modal, { dataSet: this.cache("modalDataSet", { class: "table-header-select-modal" }), onRequestClose: this.tt.onModalRequestClose, transparent: true, children: this.p.preparedColumns.map(({ column, tableSettingColumn }) => _jsx(View, { style: this.cache("headerColumnContentViewStyle", { marginVertical: 5 }), children: _jsx(HeaderColumnContent, { column: column, table: table, tableSettingColumn: tableSettingColumn, sortLinkProps: this.cache("sortLinkProps", { usePressable: true }) }) }, tableSettingColumn.identifier())) }), _jsx(Pressable, { onPress: this.tt.onSortedByPress, children: _jsxs(Text, { children: [sortedByPreparedColumn && sortMode == "asc" &&
                                t(".sorted_by_column", { column: columnLabel, defaultValue: "Sorted by %{column}" }), sortedByPreparedColumn && sortMode == "desc" &&
                                t(".sorted_by_column_reversed", { column: columnLabel, defaultValue: "Sorted by %{column} reversed" }), !sortedByPreparedColumn &&
                                t(".not_sorted", { defaultValue: "Not sorted" })] }) })] }));
    }
    onModalRequestClose = () => this.setState({ modalOpen: false });
    onSortedByPress = () => this.setState({ modalOpen: true });
}));
