// shakapacker/package/rspack/index.js
/* eslint import/no-dynamic-require: 0 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.inliningCss = exports.canProcess = exports.moduleExists = exports.rules = exports.env = exports.baseConfig = exports.generateRspackConfig = exports.devServer = exports.config = exports.unique = exports.mergeWithRules = exports.mergeWithCustomize = exports.merge = void 0;
// Mixed require/import syntax:
// - Using require() for compiled JS modules that may not have proper ES module exports
// - Using import for type-only imports and Node.js built-in modules
const path_1 = require("path");
const fs_1 = require("fs");
const webpackMerge = require("webpack-merge");
const config = require("../config");
exports.config = config;
const baseConfig = require("../environments/base");
exports.baseConfig = baseConfig;
const devServer = require("../dev_server");
exports.devServer = devServer;
const env = require("../env");
exports.env = env;
const { moduleExists, canProcess } = require("../utils/helpers");
exports.moduleExists = moduleExists;
exports.canProcess = canProcess;
const inliningCss = require("../utils/inliningCss");
exports.inliningCss = inliningCss;
const { getPlugins } = require("../plugins/rspack");
const { getOptimization } = require("../optimization/rspack");
const { validateRspackDependencies } = require("../utils/validateDependencies");
const rulesPath = (0, path_1.resolve)(__dirname, "../rules", "rspack.js");
const rules = require(rulesPath);
exports.rules = rules;
const generateRspackConfig = (extraConfig = {}, ...extraArgs) => {
    // Validate required dependencies first
    validateRspackDependencies();
    if (extraArgs.length > 0) {
        throw new Error("Only one extra config may be passed here - use webpack-merge to merge configs before passing them to Shakapacker");
    }
    const { nodeEnv } = env;
    const path = (0, path_1.resolve)(__dirname, "../environments", `${nodeEnv}.js`);
    const environmentConfig = (0, fs_1.existsSync)(path) ? require(path) : baseConfig;
    // Create base rspack config
    const rspackConfig = {
        ...environmentConfig,
        module: {
            rules
        },
        plugins: getPlugins(),
        optimization: getOptimization()
    };
    return webpackMerge.merge({}, rspackConfig, extraConfig);
};
exports.generateRspackConfig = generateRspackConfig;
// Re-export webpack-merge utilities for backward compatibility
var webpack_merge_1 = require("webpack-merge");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return webpack_merge_1.merge; } });
Object.defineProperty(exports, "mergeWithCustomize", { enumerable: true, get: function () { return webpack_merge_1.mergeWithCustomize; } });
Object.defineProperty(exports, "mergeWithRules", { enumerable: true, get: function () { return webpack_merge_1.mergeWithRules; } });
Object.defineProperty(exports, "unique", { enumerable: true, get: function () { return webpack_merge_1.unique; } });
