// @kaspernj/api-maker/build/utils/checkboxes.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable implicit-arrow-linebreak, newline-per-chained-call, no-throw-literal, sort-imports */
import { shapeComponent, ShapeComponent } from "set-state-compare/build/shape-component.js";
import React, { useMemo } from "react";
import Checkbox from "./checkbox";
import { digs } from "diggerize";
import { useForm } from "../form";
import * as inflection from "inflection";
import InvalidFeedback from "./invalid-feedback";
import memo from "set-state-compare/build/memo.js";
import PropTypes from "prop-types";
import propTypesExact from "prop-types-exact";
import Text from "./text";
import useInput from "../use-input.js";
import { View } from "react-native";
const OptionElement = memo(shapeComponent(class OptionElement extends ShapeComponent {
    static propTypes = propTypesExact({
        checked: PropTypes.bool.isRequired,
        inputName: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
        option: PropTypes.array.isRequired
    });
    render() {
        const { checked, inputName, option } = this.p;
        const dataSet = useMemo(() => ({
            component: "api-maker/utils/checkboxes/option",
            name: inputName,
            value: option[1]
        }), [inputName, option[1]]);
        return (_jsx(View, { children: _jsx(Checkbox, { checked: checked, dataSet: dataSet, label: option[0], onCheckedChange: this.tt.onChange }) }));
    }
    onChange = (checked) => this.p.onChange({ checked, option: this.p.option });
}));
export default memo(shapeComponent(class ApiMakerUtilsCheckboxes extends ShapeComponent {
    static propTypes = propTypesExact({
        attribute: PropTypes.string,
        defaultValue: PropTypes.array,
        label: PropTypes.string,
        labelClassName: PropTypes.string,
        model: PropTypes.object,
        name: PropTypes.string,
        onChange: PropTypes.func,
        options: PropTypes.array.isRequired
    });
    setup() {
        const { inputProps, wrapperOpts } = useInput({ props: this.props });
        this.generatedId = useMemo(() => Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15), []);
        this.setInstance({
            form: useForm(),
            inputProps,
            wrapperOpts
        });
        this.useStates({
            checkedOptions: () => {
                if (Array.isArray(this.props.defaultValue))
                    return this.props.defaultValue;
                if (this.props.defaultValue)
                    return [this.props.defaultValue];
                return [];
            }
        });
        useMemo(() => {
            if (this.form && inputProps.name) {
                this.form.setValue(inputProps.name, this.s.checkedOptions);
            }
        }, []);
    }
    render() {
        const { wrapperOpts } = this.tt;
        const { errors } = digs(wrapperOpts, "errors");
        return (_jsxs(View, { dataSet: this.cache("rootViewDataSet", { component: "api-maker/utils/checkboxes" }), children: [_jsx(Text, { style: this.cache("textStyle", { fontWeight: "bold" }), children: this.tt.wrapperOpts.label }), this.props.options.map((option) => _jsx(OptionElement, { checked: this.isChecked(option), inputName: this.inputName(), onChange: this.tt.onOptionChecked, option: option }, option[1])), errors.length > 0 &&
                    _jsx(InvalidFeedback, { errors: errors })] }));
    }
    inputDefaultValue() {
        const { attribute, defaultValue, model } = this.props;
        if (defaultValue) {
            return defaultValue;
        }
        else if (attribute && model) {
            if (!model[attribute])
                throw `No such attribute: ${attribute}`;
            return this.props.model[attribute]();
        }
    }
    inputName() {
        if (this.props.name) {
            return `${this.props.name}[]`;
        }
        else if (this.props.model) {
            return `${this.props.model.modelClassData().paramKey}[${inflection.underscore(this.props.attribute)}]`;
        }
    }
    isChecked = (option) => this.s.checkedOptions.includes(option[1]);
    onOptionChecked = ({ checked, option }) => {
        const { inputProps, form } = this.tt;
        const { name } = inputProps;
        let newOptions;
        if (checked) {
            newOptions = this.s.checkedOptions.concat([option[1]]);
            this.setState({ checkedOptions: newOptions });
        }
        else {
            newOptions = this.s.checkedOptions.filter((value) => value != option[1]);
            this.setState({ checkedOptions: newOptions });
        }
        if (this.props.onChange) {
            this.p.onChange({ checked, option });
        }
        if (form && name) {
            form.setValue(name, newOptions);
        }
    };
}));
