// @kaspernj/api-maker/build/utils/modal.js
import { jsx as _jsx } from "react/jsx-runtime";
/* eslint-disable implicit-arrow-linebreak, prefer-object-spread, sort-vars */
import { Modal, Pressable, View } from "react-native";
import React, { useMemo } from "react";
import BaseComponent from "../base-component";
import Card from "./card";
import Icon from "../utils/icon";
import memo from "set-state-compare/build/memo.js";
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import { useBreakpoint } from "responsive-breakpoints";
export default memo(shapeComponent(class ApiMakerUtilsComponent extends BaseComponent {
    render() {
        const { xs, sm } = useBreakpoint();
        const { children, dataSet, ...restProps } = this.props;
        const actualDataSet = useMemo(() => Object.assign({ component: "api-maker/utils/modal" }, dataSet), [dataSet]);
        let width, maxWidth;
        if (xs || sm) {
            width = "95%";
        }
        else {
            width = "80%";
            maxWidth = 800;
        }
        const cardStyle = useMemo(() => ({ width, maxWidth }), [width, maxWidth]);
        return (_jsx(Modal, { dataSet: actualDataSet, ...restProps, children: _jsx(View, { style: this.cache("rootViewStyle", {
                    alignItems: "center",
                    justifyContent: "center",
                    width: "100%",
                    height: "100%",
                    backgroundColor: "rgba(0, 0, 0, 0.8)"
                }), children: _jsx(Card, { controls: this.cardHeaderControls(), style: cardStyle, children: children }) }) }));
    }
    cardHeaderControls() {
        return (_jsx(Pressable, { onPress: this.tt.onModalClosePress, style: this.cache("pressableStyle", { marginLeft: "auto", padding: 5 }), children: _jsx(Icon, { name: "remove" }) }));
    }
    onModalClosePress = () => this.props.onRequestClose && this.props.onRequestClose();
}));
