// shakapacker/package/index.js
/* eslint import/no-dynamic-require: 0 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const webpackMerge = __importStar(require("webpack-merge"));
const path_1 = require("path");
const fs_1 = require("fs");
const config_1 = __importDefault(require("./config"));
const base_1 = __importDefault(require("./environments/base"));
const dev_server_1 = __importDefault(require("./dev_server"));
const env_1 = __importDefault(require("./env"));
const helpers_1 = require("./utils/helpers");
const inliningCss_1 = __importDefault(require("./utils/inliningCss"));
const rulesPath = (0, path_1.resolve)(__dirname, "rules", `${config_1.default.assets_bundler}.js`);
/** Array of webpack/rspack loader rules */
const rules = require(rulesPath);
/**
 * Generate webpack configuration with optional custom config.
 *
 * @param extraConfig - Optional webpack configuration to merge with base config
 * @returns Final webpack configuration
 * @throws {Error} If more than one argument is provided
 */
const generateWebpackConfig = (extraConfig = {}, ...extraArgs) => {
    if (extraArgs.length > 0) {
        throw new Error(`Invalid usage: generateWebpackConfig() accepts only one configuration object.\n\n` +
            `You passed ${extraArgs.length + 1} arguments. Only one extra config may be passed here - use webpack-merge to merge configs before passing them to Shakapacker.\n\n` +
            `Example:\n` +
            `  const { merge } = require('webpack-merge')\n` +
            `  const mergedConfig = merge(config1, config2, config3)\n` +
            `  const finalConfig = generateWebpackConfig(mergedConfig)\n\n` +
            `Or if using ES6:\n` +
            `  import { merge } from 'webpack-merge'\n` +
            `  const finalConfig = generateWebpackConfig(merge(config1, config2))`);
    }
    const { nodeEnv } = env_1.default;
    const path = (0, path_1.resolve)(__dirname, "environments", `${nodeEnv}.js`);
    const environmentConfig = (0, fs_1.existsSync)(path) ? require(path) : base_1.default;
    return webpackMerge.merge({}, environmentConfig, extraConfig);
};
module.exports = {
    /** Shakapacker configuration from shakapacker.yml */
    config: config_1.default,
    /** Development server configuration */
    devServer: dev_server_1.default,
    /** Generate webpack configuration with optional custom config */
    generateWebpackConfig,
    /** Base webpack/rspack configuration */
    baseConfig: base_1.default,
    /** Environment configuration (railsEnv, nodeEnv, etc.) */
    env: env_1.default,
    /** Array of webpack/rspack loader rules */
    rules,
    /** Check if a module exists in node_modules */
    moduleExists: helpers_1.moduleExists,
    /** Process a file if a specific loader is available */
    canProcess: helpers_1.canProcess,
    /** Whether CSS should be inlined (dev server with HMR) */
    inliningCss: inliningCss_1.default,
    /** webpack-merge functions (merge, mergeWithCustomize, mergeWithRules, unique) */
    ...webpackMerge
};
