// shakapacker/package/configExporter/fileWriter.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileWriter = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const os_1 = require("os");
/**
 * Handles writing config exports to files.
 * Supports single file output or multiple files (one per config).
 */
class FileWriter {
    /**
     * Write multiple config files (one per config in array)
     */
    static writeMultipleFiles(outputs, targetDir) {
        // Ensure directory exists
        FileWriter.ensureDirectory(targetDir);
        // Write each file
        outputs.forEach((output) => {
            const safeName = (0, path_1.basename)(output.filename);
            const filePath = (0, path_1.resolve)(targetDir, safeName);
            FileWriter.validateOutputPath(filePath);
            FileWriter.writeFile(filePath, output.content);
            console.log(`[Config Exporter] Created: ${filePath}`);
        });
        console.log(`[Config Exporter] Exported ${outputs.length} config file(s) to ${targetDir}`);
    }
    /**
     * Write a single file
     */
    static writeSingleFile(filePath, content) {
        // Ensure parent directory exists
        const dir = (0, path_1.dirname)(filePath);
        FileWriter.ensureDirectory(dir);
        FileWriter.validateOutputPath(filePath);
        FileWriter.writeFile(filePath, content);
        console.log(`[Config Exporter] Created: ${filePath}`);
    }
    /**
     * Generate filename for a config export
     * Format without build: {bundler}-{env}-{type}.{ext}
     * Format with build: {bundler}-{build}-{type}.{ext}
     *
     * @param bundler - The bundler type (webpack, rspack)
     * @param env - The environment (development, production, test)
     * @param configType - Type of config. Built-in: "client", "server", "all", "client-hmr". Custom: any string from outputs array
     * @param format - Output format (yaml, json, inspect)
     * @param buildName - Optional build name that overrides env in filename
     *
     * @example
     * // Built-in types
     * generateFilename("webpack", "development", "client", "yaml")
     * // => "webpack-development-client.yml"
     *
     * @example
     * // Custom output names
     * generateFilename("webpack", "development", "client-modern", "yaml", "dev-hmr")
     * // => "webpack-dev-hmr-client-modern.yml"
     */
    static generateFilename(bundler, env, configType, format, buildName) {
        let ext;
        if (format === "yaml") {
            ext = "yml";
        }
        else if (format === "json") {
            ext = "json";
        }
        else {
            ext = "txt";
        }
        const name = buildName || env;
        return `${bundler}-${name}-${configType}.${ext}`;
    }
    static writeFile(filePath, content) {
        (0, fs_1.writeFileSync)(filePath, content, "utf8");
    }
    static ensureDirectory(dir) {
        if (!(0, fs_1.existsSync)(dir)) {
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        }
    }
    /**
     * Validate output path and warn if writing outside cwd
     */
    static validateOutputPath(outputPath) {
        const absPath = (0, path_1.resolve)(outputPath);
        const cwd = process.cwd();
        const isWithin = (base, target) => {
            const rel = (0, path_1.relative)(base, target);
            return rel === "" || (!rel.startsWith("..") && !(0, path_1.isAbsolute)(rel));
        };
        if (!isWithin(cwd, absPath) && !isWithin((0, os_1.tmpdir)(), absPath)) {
            console.warn(`[Config Exporter] Warning: Writing to ${absPath} which is outside current directory (${cwd}) or temp (${(0, os_1.tmpdir)()})`);
        }
    }
}
exports.FileWriter = FileWriter;
