// shakapacker/package/configExporter/types.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONFIG_FILE = exports.DEFAULT_EXPORT_DIR = exports.DANGEROUS_ENV_VARS = exports.BUILD_ENV_VARS = void 0;
exports.isBuildEnvVar = isBuildEnvVar;
exports.isDangerousEnvVar = isDangerousEnvVar;
/**
 * Environment variable names that can be set by build configurations.
 * These are the only environment variables that build configs are allowed to set.
 * This whitelist prevents malicious configs from modifying critical system variables.
 */
exports.BUILD_ENV_VARS = [
    "NODE_ENV",
    "RAILS_ENV",
    "NODE_OPTIONS",
    "BABEL_ENV",
    "WEBPACK_SERVE",
    "CLIENT_BUNDLE_ONLY",
    "SERVER_BUNDLE_ONLY"
];
/**
 * Environment variables that must never be set by build configurations.
 * Setting these could compromise system security or cause unexpected behavior.
 */
exports.DANGEROUS_ENV_VARS = [
    "PATH",
    "HOME",
    "LD_PRELOAD",
    "LD_LIBRARY_PATH",
    "DYLD_LIBRARY_PATH",
    "DYLD_INSERT_LIBRARIES"
];
/**
 * Type predicate to check if a string is in the BUILD_ENV_VARS whitelist
 *
 * Note: The type assertion is necessary because TypeScript's type system cannot
 * infer that .includes() on a readonly const array will properly narrow the type.
 * The assertion is safe because we're only widening the type for the includes() check.
 */
function isBuildEnvVar(key) {
    return exports.BUILD_ENV_VARS.includes(key);
}
/**
 * Type predicate to check if a string is in the DANGEROUS_ENV_VARS blacklist
 *
 * Note: The type assertion is necessary because TypeScript's type system cannot
 * infer that .includes() on a readonly const array will properly narrow the type.
 * The assertion is safe because we're only widening the type for the includes() check.
 */
function isDangerousEnvVar(key) {
    return exports.DANGEROUS_ENV_VARS.includes(key);
}
/**
 * Default directory for config exports when using --doctor or file output modes.
 */
exports.DEFAULT_EXPORT_DIR = "shakapacker-config-exports";
/**
 * Default config file path for bundler build configurations.
 */
exports.DEFAULT_CONFIG_FILE = "config/shakapacker-builds.yml";
