// @kaspernj/api-maker/build/bootstrap/paginate.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable no-return-assign */
import React, { useMemo } from "react";
import { shapeComponent } from "set-state-compare/build/shape-component.js";
import BaseComponent from "../base-component"; // eslint-disable-line sort-imports
import Link from "../link";
import PropTypes from "prop-types";
import Result from "../result.js";
import Text from "../utils/text";
import instanceOfClassName from "../instance-of-class-name.js";
import memo from "set-state-compare/build/memo.js";
import propTypesExact from "prop-types-exact";
import qs from "qs";
import urlEncode from "../url-encode.js";
const dataSets = {};
const FIRST_PAGE_LABEL = "⇤";
const PREVIOUS_PAGE_LABEL = "←";
const NEXT_PAGE_LABEL = "→";
const LAST_PAGE_LABEL = "⇥";
const ELLIPSIS_LABEL = "…";
export default memo(shapeComponent(class ApiMakerBootstrapPaginate extends BaseComponent {
    static propTypes = propTypesExact({
        result: PropTypes.oneOfType([
            instanceOfClassName("ApiMakerResult"),
            PropTypes.instanceOf(Result)
        ]).isRequired
    });
    setup() {
        const { result } = this.p;
        this.totalPages = useMemo(() => Math.ceil(result.count() / result.perPage()), [result.count(), result.perPage()]);
        this.pages = useMemo(() => this.calculatePages(), [result.currentPage(), this.totalPages]);
    }
    calculatePages() {
        const { result } = this.p;
        const { totalPages } = this.tt;
        const currentPage = result.currentPage();
        const pages = [];
        let pagesFrom = currentPage - 5;
        let pagesTo = currentPage + 5;
        if (pagesFrom < 1) {
            pagesFrom = 1;
        }
        if (pagesTo > totalPages) {
            pagesTo = totalPages;
        }
        for (let i = pagesFrom; i <= pagesTo; i++) {
            pages.push(i);
        }
        return pages;
    }
    isPageActiveClass = (pageNumber) => {
        if (this.p.result.currentPage() == pageNumber) {
            return "active";
        }
        return "not-active";
    };
    pagePath(pageNumber) {
        let pageKey = this.p.result.data.collection.queryArgs.pageKey;
        if (!pageKey) {
            pageKey = "page";
        }
        const currentParams = qs.parse(globalThis.location.search.substr(1));
        currentParams[pageKey] = pageNumber;
        const newParams = qs.stringify(currentParams, { encoder: urlEncode });
        const newPath = `${location.pathname}?${newParams}`;
        return newPath;
    }
    previousPagePath() {
        const { result } = this.p;
        let previousPage;
        if (result.currentPage() > 1) {
            previousPage = result.currentPage() - 1;
        }
        else {
            previousPage = result.currentPage();
        }
        return this.pagePath(previousPage);
    }
    nextPagePath() {
        const { result } = this.p;
        let nextPage;
        if (result.currentPage() < this.tt.totalPages) {
            nextPage = result.currentPage() + 1;
        }
        else {
            nextPage = result.currentPage();
        }
        return this.pagePath(nextPage);
    }
    showBackwardsDots = () => this.p.result.currentPage() - 5 > 1;
    showForwardsDots = () => this.p.result.currentPage() + 5 < this.tt.totalPages;
    render() {
        const { result } = this.p;
        const { pages, totalPages } = this.tt;
        const showNextPage = result.currentPage() < totalPages;
        const showLastPage = result.currentPage() < totalPages;
        const showPreviousPage = result.currentPage() > 1;
        const showFirstPage = result.currentPage() > 1;
        const previousPageDisabled = !showPreviousPage;
        const firstPageDisabled = !showFirstPage;
        const nextPageDisabled = !showNextPage;
        const lastPageDisabled = !showLastPage;
        return (_jsxs("ul", { className: "pagination", "data-pages-length": pages.length, children: [_jsxs("li", { className: `page-item ${previousPageDisabled ? "disabled" : ""}`, children: [previousPageDisabled &&
                            _jsx(Text, { children: FIRST_PAGE_LABEL }), showPreviousPage &&
                            _jsx(Link, { dataSet: dataSets.pageLink ||= { class: "page-link" }, to: this.pagePath(1), children: _jsx(Text, { children: FIRST_PAGE_LABEL }) })] }, "page-first"), _jsxs("li", { className: `page-item ${firstPageDisabled ? "disabled" : ""}`, children: [firstPageDisabled &&
                            _jsx(Text, { children: PREVIOUS_PAGE_LABEL }), showFirstPage &&
                            _jsx(Link, { dataSet: dataSets.pageLink ||= { class: "page-link" }, to: this.previousPagePath(), children: _jsx(Text, { children: PREVIOUS_PAGE_LABEL }) })] }, "page-previous"), this.showBackwardsDots() &&
                    _jsx("li", { className: "page-item disabled", children: _jsx(Text, { children: ELLIPSIS_LABEL }) }), pages.map((page) => ( // eslint-disable-line no-extra-parens
                _jsxs("li", { className: `page-item ${this.isPageActiveClass(page)}`, "data-active": this.isPageActiveClass(page) == "active", "data-page": page, children: [this.isPageActiveClass(page) == "active" &&
                            _jsx(Text, { children: page }), this.isPageActiveClass(page) == "not-active" &&
                            _jsx(Link, { dataSet: dataSets.pageLink ||= { class: "page-link" }, to: this.pagePath(page), children: _jsx(Text, { children: page }) })] }, `page-${page}`))), this.showForwardsDots() &&
                    _jsx("li", { className: "page-item disabled", children: _jsx(Text, { children: ELLIPSIS_LABEL }) }), _jsxs("li", { className: `page-item ${nextPageDisabled ? "disabled" : ""}`, children: [nextPageDisabled &&
                            _jsx(Text, { children: NEXT_PAGE_LABEL }), showNextPage &&
                            _jsx(Link, { dataSet: dataSets.pageLink ||= { class: "page-link" }, to: this.nextPagePath(), children: _jsx(Text, { children: NEXT_PAGE_LABEL }) })] }, "page-next"), _jsxs("li", { className: `page-item ${lastPageDisabled ? "disabled" : ""}`, children: [lastPageDisabled &&
                            _jsx(Text, { children: LAST_PAGE_LABEL }), showLastPage &&
                            _jsx(Link, { dataSet: dataSets.pageLink ||= { class: "page-link" }, to: this.pagePath(totalPages), children: _jsx(Text, { children: LAST_PAGE_LABEL }) })] }, "page-last")] }));
    }
}));
